File: hub/sdk/python-client/src/central_auth_hub/token_exchange.py
"""Token exchange — POST /oauth/token."""

from __future__ import annotations

import httpx

from .config import Config
from .errors import TokenError


def exchange_code(
    config: Config, token_endpoint: str, code: str, code_verifier: str
) -> dict:
    """Exchange auth code → access_token (server-to-server, sync)."""
    try:
        r = httpx.post(
            token_endpoint,
            data={
                "grant_type": "authorization_code",
                "code": code,
                "client_id": config.client_id,
                "client_secret": config.client_secret,
                "redirect_uri": config.redirect_uri,
                "code_verifier": code_verifier,
            },
            headers={"Accept": "application/json"},
            timeout=config.http_timeout,
        )
    except Exception as e:
        raise TokenError(f"Token endpoint unreachable: {e}") from e

    if r.status_code != 200:
        raise TokenError(f"Token exchange HTTP {r.status_code}: {r.text[:200]}")
    data = r.json()
    if "access_token" not in data:
        raise TokenError("Response missing access_token")
    return data


async def exchange_code_async(
    config: Config, token_endpoint: str, code: str, code_verifier: str
) -> dict:
    """Async variant."""
    try:
        async with httpx.AsyncClient(timeout=config.http_timeout) as client:
            r = await client.post(
                token_endpoint,
                data={
                    "grant_type": "authorization_code",
                    "code": code,
                    "client_id": config.client_id,
                    "client_secret": config.client_secret,
                    "redirect_uri": config.redirect_uri,
                    "code_verifier": code_verifier,
                },
                headers={"Accept": "application/json"},
            )
    except Exception as e:
        raise TokenError(f"Token endpoint unreachable: {e}") from e
    if r.status_code != 200:
        raise TokenError(f"Token exchange HTTP {r.status_code}: {r.text[:200]}")
    data = r.json()
    if "access_token" not in data:
        raise TokenError("Response missing access_token")
    return data

File: hub/sdk/node-client/src/tokenExchange.ts
import { Config } from "./config.js";
import { TokenError } from "./errors.js";

export interface TokenResponse {
  access_token: string;
  token_type?: string;
  expires_in?: number;
  scope?: string;
  [k: string]: unknown;
}

/** Exchange auth code → access_token via POST /oauth/token. */
export async function exchangeCode(
  config: Config,
  tokenEndpoint: string,
  code: string,
  codeVerifier: string
): Promise<TokenResponse> {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), config.httpTimeout);
  try {
    const body = new URLSearchParams({
      grant_type: "authorization_code",
      code,
      client_id: config.clientId,
      client_secret: config.clientSecret,
      redirect_uri: config.redirectUri,
      code_verifier: codeVerifier,
    });
    const r = await fetch(tokenEndpoint, {
      method: "POST",
      signal: ctrl.signal,
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json",
      },
      body,
    });
    if (!r.ok) {
      const txt = await r.text();
      throw new TokenError(`Token exchange HTTP ${r.status}: ${txt.slice(0, 200)}`);
    }
    const data = (await r.json()) as TokenResponse;
    if (!data.access_token) throw new TokenError("Response missing access_token");
    return data;
  } catch (e) {
    if (e instanceof TokenError) throw e;
    throw new TokenError(`Token endpoint unreachable: ${(e as Error).message}`);
  } finally {
    clearTimeout(t);
  }
}
